File: transform/keywords-scala/src/main/scala/bigdt/transform/keywords/TrendingWordsJob.scala
package bigdt.transform.keywords

import org.apache.spark.sql.{SparkSession, DataFrame}
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types.ArrayType
import scala.collection.mutable
import scala.collection.JavaConverters._
import scala.io.Source
import vn.pipeline.{VnCoreNLP, Annotation}
import java.io.File

/**
 * BigDT Transform - Trending Keywords Extraction
 * Scala single-file implementation: VnCoreNLP-based Vietnamese NLP + Spark aggregation
 */
object TrendingWordsJob {
  private val REGEX_TEXT = """[^a-zA-Z0-9_àáảãạăằắẳẵặâầấẩẫậèéẻẽẹêềếểễệìíỉĩịòóỏõọôồốổỗộơờớởỡợùúủũụưừứửữựỳýỷỹỵđ\s]"""
  private val DEFAULT_HDFS_BASE = "hdfs://namenode:9000/raw_zone"
  private val DEFAULT_HDFS_OUTPUT = "hdfs://namenode:9000/work_zone/table_trending_words"

  private val stopwordsCache = mutable.Map[String, Set[String]]()
  @transient private lazy val vnPipeline = {
    try {
      Some(new VnCoreNLP(Array("wseg", "pos")))
    } catch {
      case _: Exception => None
    }
  }

  def main(args: Array[String]): Unit = {
    var hdfsBase = sys.env.getOrElse("HDFS_BASE_PATH", DEFAULT_HDFS_BASE)
    var hdfsOutput = sys.env.getOrElse("HDFS_OUTPUT_PATH", DEFAULT_HDFS_OUTPUT)
    if (args.length > 0) hdfsBase = args(0)
    if (args.length > 1) hdfsOutput = args(1)
    runJob(hdfsBase, hdfsOutput)
  }

  private def loadStopwords(): Set[String] = {
    stopwordsCache.getOrElseUpdate("vietnamese", {
      try {
        val resource = getClass.getResourceAsStream("/stopwords_vi.txt")
        if (resource != null) {
          val source = Source.fromInputStream(resource, "UTF-8")
          val stopwords = source.getLines().map(_.trim).filter(_.nonEmpty).toSet
          source.close()
          stopwords
        } else Set.empty[String]
      } catch {
        case _: Exception => Set.empty[String]
      }
    })
  }

  private def extractKeywords(text: String, minWordLength: Int = 2): Array[String] = {
    if (text == null || text.trim.isEmpty) return Array()
    val stopwords = loadStopwords()
    vnPipeline match {
      case Some(pipeline) =>
        try {
          val annotation = new Annotation(text)
          pipeline.annotate(annotation)
          val keywords = mutable.ArrayBuffer[String]()
          
          annotation.getSentences.asScala.foreach { sentence =>
            val words = sentence.getWords.asScala.toList
            
            // Get properly segmented words from VnCoreNLP
            val forms = words.map(w => w.getForm.toLowerCase)
            val posTags = words.map(w => {
              val tag = w.getPosTag
              if (tag != null && tag.nonEmpty) tag.substring(0, 1) else "X"
            })
            
            // Extract noun phrases: N+ (consecutive nouns/adjectives)
            var i = 0
            while (i < forms.length) {
              val pos = posTags(i)
              if (pos == "N" || pos == "A") {
                val phrase = mutable.ArrayBuffer[String]()
                phrase += forms(i)
                var j = i + 1
                while (j < forms.length && (posTags(j) == "N" || posTags(j) == "A")) {
                  phrase += forms(j)
                  j += 1
                }
                val phraseStr = phrase.mkString(" ")
                if (phraseStr.length >= minWordLength && !stopwords.contains(phraseStr))
                  keywords += phraseStr
                i = j - 1
              }
              i += 1
            }
            
            // Extract verb phrases: V+ followed by optional N/A
            i = 0
            while (i < forms.length) {
              val pos = posTags(i)
              if (pos == "V") {
                val phrase = mutable.ArrayBuffer[String]()
                phrase += forms(i)
                var j = i + 1
                while (j < forms.length && posTags(j) == "V") {
                  phrase += forms(j)
                  j += 1
                }
                if (j < forms.length && (posTags(j) == "N" || posTags(j) == "A")) {
                  phrase += forms(j)
                  j += 1
                }
                val phraseStr = phrase.mkString(" ")
                if (phraseStr.length >= minWordLength && !stopwords.contains(phraseStr))
                  keywords += phraseStr
                i = j - 1
              }
              i += 1
            }
            
            // Add individual non-stopword tokens with VnCoreNLP segmentation
            forms.zip(posTags).foreach { case (form, pos) =>
              if (form.length >= minWordLength && !stopwords.contains(form) && (pos == "N" || pos == "A" || pos == "V"))
                keywords += form
            }
          }
          
          if (keywords.nonEmpty) keywords.distinct.toArray else simpleTokenize(text, stopwords, minWordLength)
        } catch {
          case _: Exception => simpleTokenize(text, stopwords, minWordLength)
        }
      case None => simpleTokenize(text, stopwords, minWordLength)
    }
  }

  private def simpleTokenize(text: String, stopwords: Set[String], minWordLength: Int): Array[String] = {
    val tokens = text.toLowerCase.trim.split("\\s+").filter(_.nonEmpty)
    val filtered = tokens.filter(w => w.length >= minWordLength && !stopwords.contains(w))
    val keywords = mutable.ArrayBuffer[String]()
    for (len <- 4 to 2 by -1; i <- 0 to (filtered.length - len))
      keywords += filtered.slice(i, i + len).mkString(" ")
    (keywords ++ filtered).distinct.toArray
  }

  def runJob(hdfsBase: String, hdfsOutput: String): Unit = {
    val spark = SparkSession
      .builder()
      .appName("extraction_job:v1.0.0")
      .config("spark.executor.memory", "2g")
      .config("spark.executor.cores", "2")
      .config("spark.driver.memory", "2g")
      .getOrCreate()

    try {
      println("Starting trending words extraction...")
      val categories = Map(
        "giai_tri" -> "GiaiTri",
        "cong_nghe" -> "CongNghe",
        "suc_khoe" -> "SucKhoe"
      )

      val dfs = categories.flatMap { case (folder, catName) =>
        try {
          val path = s"${hdfsBase.stripSuffix("/")}/$folder"
          val df = spark.read
            .option("inferSchema", "true")
            .json(path)
            .withColumn("chu_de", lit(catName))
          Some(df)
        } catch {
          case e: Exception =>
            println(s"[WARN] Failed to load $folder: ${e.getMessage}")
            None
        }
      }.toSeq

      if (dfs.isEmpty) throw new Exception("No data found in any category folder!")

      var df = dfs.reduce(_ unionByName _)
      println(s"Loaded ${df.count()} records")

      df = df.select(
        date_format(col("publish_date"), "yyyyMMdd").alias("ngay"),
        col("source").alias("nguon"),
        col("chu_de"),
        col("title"),
        col("content")
      )
      
      df = df.filter(col("ngay").isNotNull)

      df = df.withColumn(
        "full_text",
        regexp_replace(
          concat_ws(" ", col("title"), col("content")),
          REGEX_TEXT,
          " "
        )
      )

      val extractKeywordsUDF = udf((text: String) => extractKeywords(text))
      df = df.withColumn("keywords", extractKeywordsUDF(col("full_text")))

      df = df.select(
        col("ngay"),
        col("nguon"),
        col("chu_de"),
        explode(col("keywords")).alias("tu_khoa")
      )
      
      df = df.filter(col("tu_khoa").isNotNull && length(col("tu_khoa")) >= 2)

      val resultDF = df
        .groupBy("ngay", "nguon", "chu_de", "tu_khoa")
        .count()
        .withColumnRenamed("count", "so_lan_xuat_hien")
        .orderBy(desc("so_lan_xuat_hien"))

      resultDF.write.mode("overwrite").parquet(hdfsOutput)
      resultDF.coalesce(1).write.mode("overwrite").option("header", "true").csv(s"${hdfsOutput}_csv")

      println(s"Completed: ${resultDF.count()} keywords")
      resultDF.limit(10).show(truncate = false)

    } catch {
      case e: Exception =>
        println(s"[ERROR] ${e.getMessage}")
        e.printStackTrace()
        throw e
    } finally {
      spark.stop()
    }
  }
}

File: transform/keywords/trending_words_job.py
"""BigDT Transform - Trending Keywords Extraction"""
import os, sys
from functools import reduce
from pyvi import ViTokenizer
from pyspark.sql import SparkSession, DataFrame
from pyspark.sql.functions import col, explode, regexp_replace, desc, date_format, udf, concat_ws, lit
from pyspark.sql.types import ArrayType, StringType

REGEX_TEXT = r"[^a-zA-Z0-9_àáảãạăằắẳẵặâầấẩẫậèéẻẽẹêềếểễệìíỉĩịòóỏõọôồốổỗộơờớởỡợùúủũụưừứửữựỳýỷỹỵđ\s]"
def _load_stopwords():
    stopwords = set()
    path = os.path.join(os.path.dirname(__file__), "stopwords_vi.txt")
    try:
        with open(path, "r", encoding="utf-8") as f:
            stopwords.update(line.strip() for line in f if line.strip())
    except:
        pass
    return stopwords

def tokenize(text):
    if not text or not isinstance(text, str):
        return []
    try:
        return ViTokenizer.tokenize(text).split() if ViTokenizer else text.split()
    except:
        return text.split()

def extract_keywords(text, min_length=2):
    stopwords = _load_stopwords()
    tokens = tokenize(text.lower().strip())
    return [t for t in tokens if t not in stopwords and len(t) > 1 and not t.startswith("http") and len(t) >= min_length]

'''
- Main job function to run the Spark job for trending keywords extraction
- Reads JSON files from HDFS, processes them, and writes results back to HDFS in
'''
def run_job(hdfs_base, hdfs_output):
    spark = SparkSession.builder.appName("extraction_job:v1.0.0").config("spark.executor.memory", "2g").config("spark.executor.cores", "2").config("spark.driver.memory", "2g").getOrCreate()
    
    try:
        print("Starting trending words extraction...")
        categories = {'giai_tri': 'GiaiTri', 'cong_nghe': 'CongNghe', 'suc_khoe': 'SucKhoe'}
        dfs = []
        
        for folder, cat_name in categories.items():
            try:
                df_cat = spark.read.option("inferSchema", "true").json(f"{hdfs_base.rstrip('/')}/{folder}")
                dfs.append(df_cat.withColumn("chu_de", lit(cat_name)))
            except:
                pass
        
        if not dfs:
            raise Exception("No data found!")
        
        df = reduce(DataFrame.unionByName, dfs)
        print(f"Loaded {df.count()} records")
        
        df = df.select(
            date_format(col("publish_date"), "yyyyMMdd").alias("ngay"),
            col("source").alias("nguon"),
            col("chu_de"),
            col("title"),
            col("content")
        ).filter(col("ngay").isNotNull())
        
        df = df.withColumn("full_text",regexp_replace(concat_ws(" ", col("title"), col("content")),REGEX_TEXT," "))
        df = df.withColumn("keywords", udf(extract_keywords, ArrayType(StringType()))(col("full_text")))
        df = df.select(
            col("ngay"), col("nguon"), col("chu_de"), 
            explode(col("keywords")).alias("tu_khoa")
        ).filter(col("tu_khoa").isNotNull())
        
        result_df = df.groupBy("ngay", "nguon", "chu_de", "tu_khoa").count().withColumnRenamed("count", "so_lan_xuat_hien").orderBy(desc("so_lan_xuat_hien"))
        result_df.write.mode("overwrite").parquet(hdfs_output)
        result_df.coalesce(1).write.mode("overwrite").option("header", "true").csv(f"{hdfs_output}_csv")
        
        print(f"✓ Completed: {result_df.count()} keywords")
        result_df.limit(10).show(truncate=False)
        
    except Exception as e:
        print(f"[ERROR] {e}")
        raise
    finally:
        spark.stop()

if __name__ == "__main__":
    hdfs_base = os.getenv("HDFS_BASE_PATH", "hdfs://localhost:9870/raw_zone")
    hdfs_output = os.getenv("HDFS_OUTPUT_PATH", "hdfs://localhost:9870/work_zone/table_trending_words")
    if len(sys.argv) > 1:
        hdfs_base = sys.argv[1]
    if len(sys.argv) > 2:
        hdfs_output = sys.argv[2]
    run_job(hdfs_base, hdfs_output)